src/simularium/VisData.ts

```diff
@@ -98,91 +50,58 @@ class VisData {
+        this.frameCache = new VisDataCache();
         this.frameToWaitFor = 0;
         this.lockedForFrame = false;
         this.timeStepSize = 0;
+
+        this.onError = noop;
     }
 
-    //get time() { return this.cacheFrame < this.frameDataCache.length ? this.frameDataCache[this.cacheFrame] : -1 }
-    public get currentFrameData(): FrameData {
-        if (this.frameDataCache.length > 0) {
-            if (this.cacheFrame < 0) {
-                return this.frameDataCache[0];
-            } else if (this.cacheFrame >= this.frameDataCache.length) {
-                return this.frameDataCache[this.frameDataCache.length - 1];
-            } else {
-                return this.frameDataCache[this.cacheFrame];
-            }
-        }
+    public setOnError(onError: (error: FrontEndError) => void): void {
+        this.onError = onError;
+    }
 
-        return { frameNumber: 0, time: 0 };
+    public get currentFrameData(): CachedFrame {
+        let currentData: CachedFrame = nullCachedFrame();
+        if (!this.frameCache.hasFrames()) {
+            return currentData;
+        }
+        if (this.currentFrameNumber < 0) {
+            currentData = this.frameCache.getFirstFrame();
+            this.currentFrameNumber = currentData.frameNumber;
+        } else if (
+            this.frameCache.containsFrameAtFrameNumber(this.currentFrameNumber)
+        ) {
+            currentData = this.frameCache.getFrameAtFrameNumber(
+                this.currentFrameNumber
+            );
+        }
+        return currentData;
     }
 
     /**
      *   Functions to check update
      * */
     public hasLocalCacheForTime(time: number): boolean {
-        // TODO: debug compareTimes
-        if (!this.enableCache) {
-            return false;
-        }
-        if (this.frameDataCache.length < 1) {
-            return false;
-        }
-
-        const firstFrameTime = this.frameDataCache[0].time;
-        const lastFrameTime =
-            this.frameDataCache[this.frameDataCache.length - 1].time;
-
-        const notLessThanFirstFrameTime =
-            compareTimes(time, firstFrameTime, this.timeStepSize) !== -1;
-        const notGreaterThanLastFrameTime =
-            compareTimes(time, lastFrameTime, this.timeStepSize) !== 1;
-        return notLessThanFirstFrameTime && notGreaterThanLastFrameTime;
+        return this.frameCache.containsTime(time);
     }
 
     public gotoTime(time: number): void {
-        this.cacheFrame = -1;
-
-        // Find the index of the frame that has the time matching our target time
-        const frameNumber = this.frameDataCache.findIndex((frameData) => {
-            return compareTimes(frameData.time, time, this.timeStepSize) === 0;
-        });
-
-        // frameNumber is -1 if findIndex() above doesn't find a match
-        if (frameNumber !== -1) {
-            this.cacheFrame = frameNumber;
+        const frameNumber = this.frameCache.getFrameAtTime(time)?.frameNumber;
```

Comment:
`?` checks for undefined or null, not just falsy, so 0 should not trigger it.

Happy to make this more robust, but fwiw `getFrameAtTime` just returns a call of `getFrameAtTimeOrFrameNumber` with the "time" flag provided. That function has error handling and should either return the data for the frame, or throw an error.

Reply:
The `?` is also currently not required - `getFrameAtTime`'s current return type is just `CachedFrame` because it errors out when a frame at `time` is missing. (Hopefully it's clear from the above that I'm advocating for it to become required, rather than for it to be removed.)